Sort nodename, nodeaddrs, nodehostnames when doing scontrol update

hanwen-pcluste opened the pull request:
### Description of changes
* `update_nodes` function is called from multiple places. What we care about is the callings with nodeaddrs or nodehostnames. Moreover, we only apply the sort when there is nodeaddrs or nodehostnames
* There are two calling places with nodeaddrs or nodehostnames: `reset_nodes` and `_update_slurm_node_addrs`
* In `reset_nodes`, the nodename, nodeaddrs, and nodehostnames are identical. We use logic sort  the nodename, and assign the value to nodeaddrs and nodehostnames
* In `_update_slurm_node_addrs`, we know that all the nodes and addresses are already expanded (i.e. there is not Slurm compressed list expression). We use `zip` to sort the nodeaddrs and nodehostnames according to nodename
* The sort is put after the nodes are batched because `scontrol` accepts 5000 nodes at most. Putting the sort after the nodes are batches will make sure not hitting the limit

### Tests
* Unit tests are changed accordingly.
* Manual testing
We tested 30000 nodenames, addresses, hostnames mismatch can be recovered by clustermgtd.
* Whole integration tests (without AD integration and multiple FSx integration) look ok.

### Checklist
- [ ] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [ ] Check all commits' messages are clear, describing what and why vs how.
- [ ] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

hanwen-cluster committed:
Sort nodename, nodeaddrs, nodehostnames when doing scontrol update

* `update_nodes` function is called from multiple places. What we care about is the callings with nodeaddrs or nodehostnames. Moreover, we only apply the sort when there is nodeaddrs or nodehostnames
* There are two calling places with nodeaddrs or nodehostnames: `reset_nodes` and `_update_slurm_node_addrs`
* In `reset_nodes`, the nodename, nodeaddrs, and nodehostnames are identical. We use logic sort  the nodename, and assign the value to nodeaddrs and nodehostnames
* In `_update_slurm_node_addrs`, we know that all the nodes and addresses are already expanded (i.e. there is not Slurm compressed list expression). We use `zip` to sort the nodeaddrs and nodehostnames according to nodename
* The sort is put after the nodes are batched because `scontrol` accepts 5000 nodes at most. Putting the sort after the nodes are batches will make sure not hitting the limit

chenwany reviewed (commented):
Could you add more test cases to `test_batch_node_info` with nodenames not sorted?

chenwany commented on src/common/schedulers/slurm_commands.py:
```diff
@@ -165,6 +165,14 @@ def _batch_node_info(nodenames, nodeaddrs, nodehostnames, batch_size):
         # Only split on , if there is ] before
         # For ex. "node-[1,3,4-5],node-[20,30]" should split into ["node-[1,3,4-5]","node-[20,30]"]
         nodenames = re.split("(?<=]),", nodenames)
+    if nodeaddrs and nodehostnames:
```
The input of nodeaddrs here can be a string

hanwen-pcluste replied:
I refactored the code

hanwen-pcluste commented on src/common/schedulers/slurm_commands.py:
```diff
@@ -97,6 +97,25 @@ def update_nodes(
+    # For ex. "node-[1,3,4-5],node-[20,30]" should split into ["node-[1,3,4-5]","node-[20,30]"]
+    nodes = re.split("(?<=]),", nodes) if type(nodes) is str else nodes
+    nodeaddrs = re.split("(?<=]),", nodeaddrs) if type(nodeaddrs) is str else nodeaddrs
+    nodehostnames = re.split("(?<=]),", nodehostnames) if type(nodehostnames) is str else nodehostnames
```
We need to double check if these are needed

hanwen-pcluste replied:
We do not need it

hanwen-pcluste commented on src/common/schedulers/slurm_commands.py:
```diff
@@ -202,8 +221,15 @@ def set_nodes_power_down(nodes, reason=None):
 
 def reset_nodes(nodes, state=None, reason=None, raise_on_error=False):
     """Reset nodeaddr and nodehostname to be equal to nodename."""
+    nodes_str = ",".join(nodes) if type(nodes) is list else nodes
```
We need to double check if these are needed

hanwen-pcluste replied:
We do not need it

charlesg3 commented on src/common/schedulers/slurm_commands.py:
```diff
@@ -97,6 +97,25 @@ def update_nodes(
+        # Sorting is only necessary if we set nodeaddrs or nodehostnames
+        nodes_str = ",".join(nodes)
+        sorted_node_names = (
+            check_command_output(f"scontrol show hostlistsorted {nodes_str} | xargs scontrol show hostname", shell=True)
```
Let's just put the command on a separate line which will make this easier to read

demartinofra replied:
what if this command fails? is the error handled already and the failure reason clear enough?
is there a limit to the number of nodes that these commands can handle?

jdeamicis replied:
(minor) the official command is `scontrol show hostnames`, but `hostname` works too.

hanwen-pcluste replied:
Addressed [charlesg3](https://github.com/charlesg3) and [jdeamicis](https://github.com/jdeamicis)'s comments

> what if this command fails? is the error handled already and the failure reason clear enough?
is there a limit to the number of nodes that these commands can handle?

check_command_output has try, catch and logging

demartinofra commented on src/common/schedulers/slurm_commands.py:
```diff
@@ -97,6 +97,25 @@ def update_nodes(
     if we run scontrol update state=fail_state nodeaddr=good_addr nodename=name,
     the scontrol command will fail but nodeaddr will be updated to good_addr.
     """
+    # Only split on , if there is ] before
```
please move the new code to a separate function

hanwen-pcluste replied:
Done

demartinofra commented on src/common/schedulers/slurm_commands.py:
```diff
@@ -97,6 +97,25 @@ def update_nodes(
+    # Only split on , if there is ] before
+    # For ex. "node-[1,3,4-5],node-[20,30]" should split into ["node-[1,3,4-5]","node-[20,30]"]
+    nodes = re.split("(?<=]),", nodes) if type(nodes) is str else nodes
+    nodeaddrs = re.split("(?<=]),", nodeaddrs) if type(nodeaddrs) is str else nodeaddrs
+    nodehostnames = re.split("(?<=]),", nodehostnames) if type(nodehostnames) is str else nodehostnames
```
probably you don't need a regex split for nodeaddrs and nodehostnames but simply need to split on ","

hanwen-pcluste replied:
I removed the code. My code is only executed if if nodeaddrs or hostnames are are/is being changed. The input in this case should be list already

jdeamicis committed:
Fix small bugs in new batched scontrol update logic

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

hanwen-cluster committed:
Unit tests: Fix test_batch_node_info

jdeamicis committed:
Add unit test scenario we currently cannot handle

The added unit test scenario is legit for Slurm's syntax, but
node's batching logic of input parameters for scontrol update
cannot handle strings using different grouping.

This situation should be impossible in the current logic of
clustermgtd, but in the future we should try to support any
legit combination of arguments for scontrol update.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

hanwen-cluster committed:
Minor fix of source code

The problem is discovered by unittests. Our actual code does not touch this corner case

hanwen-cluster committed:
Fix unit tests

chenwany commented on src/common/schedulers/slurm_commands.py:
```diff
@@ -104,17 +104,46 @@ def update_nodes(
 
 
+def _sort_nodes_attributes(nodes, nodeaddrs, nodehostnames):
+    nodes_str = ",".join(nodes) if type(nodes) is tuple else nodes
```
Is there any case for nodes type to be not tuple?

hanwen-pcluste replied:
It should be always tuple because it is the output of `def grouper(iterable, n):`

hanwen-cluster committed:
Remove unnecessary tuple type check

nodes in `_sort_nodes_attributes` should be always tuple because it is the output of `def grouper(iterable, n):`

hanwen-cluster committed:
Fix code-linters
